---
title: "MRG Peacock Wrasse Data Analysis"
author: "Grace"
date: "2025-04-20"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Cod Data Set

## \* Libraries

```{r libraries}
library(adegenet)
library(ade4)
library(hierfstat)
library(RColorBrewer)
library(tidyverse)
library(ggplot2)
library(egg) 
library(ggsci)
```

## 1. Estimate Genetic Diversity (per population)

### a.) Load the dataset

```{r data}
# import dataset.gen into R
x <- import2genind("cod_greenland-iceland-canada.gen", ncode = 3)
```

### b.) Split data by population

```{r}
x.pop = seppop(x)
summary.by.pop = lapply(x.pop, summary)
```

### c.) *Expected heterozygosity (He)* per population

```{r}
Hexp.ls = rep(NA, length(summary.by.pop))
for (i in 1:length(summary.by.pop)){
  Hexp.ls[i] = mean(summary.by.pop[[i]]$Hexp)
}
```

```{r plot He}
bp1 <- barplot(Hexp.ls, names.arg = levels(pop(x)), las = 2, main = "Expected heterozygosity", ylab = "He")

# Add text labels on top of bars
#text(x = bp1, y = Hexp.ls, label = round(Hexp.ls, 2), pos = 1)
```


### d.) *Observed heterozygosity (Ho)* per population

```{r}
Hobs.ls = rep(NA, length(summary.by.pop))
for (i in 1:length(summary.by.pop)){
  # Use na.rm = TRUE to ignore NA values
  Hobs.ls[i] = mean(summary.by.pop[[i]]$Hobs, na.rm = TRUE)
}
bp2 <- barplot(Hobs.ls, names.arg = levels(pop(x)), las = 2, main =
          "Observed heterozygosity", ylab = "Ho")
```

### e.) Generate a grouped bar chart to combine He and Ho


```{r}
pop_names <- levels(pop(x))

# Create a long-format data frame
he_df <- data.frame(
  Population = rep(pop_names, 2),
  Heterozygosity = c(Hexp.ls, Hobs.ls),
  Type = rep(c("Expected", "Observed"), each = length(pop_names))
)

# Shorten population names to first 3 letters
#he_df$Population <- substr(he_df$Population, 1, 3)

# Stacked barplot
ggplot(he_df, aes(x = Population, y = Heterozygosity, fill = Type)) +
  geom_bar(stat = "identity", position = "dodge") +  
  labs(title = "Expected vs Observed Heterozygosity",
       y = "Heterozygosity",
       x = "Population") +
  #geom_text(aes(label = round(Heterozygosity, 3)),  # Add numbers
            #position = position_dodge(width = 0.9), # Ensure alignment
            #vjust = -0.5,  # Adjust vertical position
            #size = 3) +  # Control text size
  theme_classic() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1),
        axis.title.x = element_text(margin = margin(t = 10)),  
        axis.title.y = element_text(margin = margin(r = 10))
        )+
  scale_fill_manual(values = pal_npg("nrc")(10)[c(4, 3)])

```

## 2. Determination of candidates under positive selection

This identifies outlier SNPs (under selection) using BayeScan output Source: BayeScan <http://cmpg.unibe.ch/software/BayeScan/versions.html>

**Plot Bayescan result in R**

### a.) Read the Bayescan results

```{r}
data <- read.table("bayes_fst_cod_cleaned.txt", header = TRUE, sep = "", stringsAsFactors = FALSE)
```


### b.) Optional: set a threshold for outliers (commonly q-value \< 0.05)

```{r}
threshold <- 0.05
```

### c.) Flag significant SNPs / Classify SNPs as Outlier or Neutral.

```{r}
data <-
  data %>%
  mutate(Significant = ifelse(qval < threshold, "Outlier", "Neutral"))
```

### d.) Plot

```{r}
ggplot(data, aes(x = -log10(qval), y = fst, color = Significant)) +
  geom_point(alpha = 0.7, size = 3) +
  scale_color_manual(values = c("Neutral" = "gray", "Outlier" = "red")) +
  theme_minimal() +
  labs(
    title = "BayeScan FST Plot",
    x = "-log10(q-value)",
    y = "FST",
    color = "Category"
  ) +
  xlim(0,5) +
  theme(
    plot.title = element_text(hjust = 0.5, face = "bold"),
    legend.position = "right"
  )+
  theme_bw() +
  scale_fill_npg()
```

### e.) save outlier snps

```{r}
outlier_snps <-
  data %>% 
  filter(Significant == "Outlier")

#I added this one: save neutral snps for later
neutral_snps <-
  data %>% 
  filter(Significant == "Neutral")
```

## 3. Subset Genotype Dataset for Outliers (and Neutral)

### a.) Import full dataset again using import2genind()

```{r}
dataset_all <- import2genind("cod_greenland-iceland-canada.gen", ncode =3)
```
### b.) subset dataset for outlier snps (and neutral snps)

```{r for outlier}
outlier_dataset <- x[loc=outlier_snps$Locus]
locNames(outlier_dataset) # Verify extracted SNPs
```
```{r for neutral}
neutral_dataset <- x[loc = neutral_snps$Locus]  # Subset dataset using correct SNP names
locNames(neutral_dataset)  # Verify extracted SNPs
```

## 4. Calculate Pairwise FST from Outliers

a.) Use a custom **pairwise.neifst()** function to compute pairwise Neiâ€™s FST between populations.

```{r}
#create the function
pairwise.neifst <- function(dat,diploid=TRUE){
  if (is.genind(dat)) dat<-genind2hierfstat(dat)
  dat<-dat[order(dat[,1]),]
  pops<-unique(dat[,1])
  npop<-length(pops)
  fstmat <- matrix(nrow=npop,ncol=npop,dimnames=list(pops,pops))
  if (is.factor(dat[,1])) {
    dat[,1]<-as.numeric(dat[,1])
    pops<-as.numeric(pops)
  }
  for(a in 2:npop){
    for(b in 1:(a-1)){
      subdat <- dat[dat[,1] == pops[a] | dat[,1]==pops[b],]
      fstmat[a,b]<-fstmat[b,a]<- basic.stats(subdat,diploid=diploid)$overall[8]
    }
    
  }
  fstmat
}
#use the function
Fst.outlier <- pairwise.neifst(outlier_dataset) 

```

b.) Visualize the FST matrix as a heatmap.

```{r}
# Shorten row and column labels to first 3 letters
#rownames(Fst.outlier) <- substr(rownames(Fst.outlier), 1, 3)
#colnames(Fst.outlier) <- substr(colnames(Fst.outlier), 1, 3)

heatmap(as.matrix(Fst.outlier), symm=T, col=brewer.pal(20,"YlGnBu"))
```

## 5. PCA Based on Outlier SNPs (Demographic Inference)

```{r library}
library(stats)
```

a.) Replace missing values (with 1) in FST matrix.

```{r}
Fst.outlier_mod <- 
  Fst.outlier %>% 
  as.data.frame() %>%
  mutate(across(everything(), ~replace_na(., mean(., na.rm = TRUE)))) %>%
  as.matrix()
```

b.) Run PCA with **princomp()** on outlier FST.

```{r}
outlier_pca <- princomp(as.matrix(Fst.outlier_mod))
```

c.) Calculate variance explained.

```{r}
variance <-outlier_pca$sdev^2 / sum(outlier_pca$sdev^2)
head(variance)
```

d.) Plot the first 2 PCA components with population labels.

```{r}
components <- as.data.frame(outlier_pca$scores[,1:6])

plot(components$Comp.1, components$Comp.2, xlab = "PC 1 - 54%",
     ylab= "PC 2 - 27%")

text(components$Comp.1, components$Comp.2, row.names(components),
     cex=0.6, pos=2)

abline(h =0, lty="dashed", untf = FALSE)
abline(v =0, lty="dashed", untf = FALSE)
```

## 6. Repeat PCA for Neutral SNPs

a.) Subset the data to get neutral SNPs (exclude outliers).

```{r}
#neutral_dataset <- x[loc=-outlier_snps$snp_pos]
#original code not working 
#Error in -outlier_snps$snp_pos : invalid argument to unary operator

#already done in the previous part
```

```{r}
Fst.neutral <- pairwise.neifst(neutral_dataset) 
```

```{r}
Fst.neutral_mod <- 
  Fst.neutral %>% 
  as.data.frame() %>%
   mutate(across(everything(), ~replace_na(., mean(., na.rm = TRUE))))  %>%
  as.matrix()
```


b.) Run PCA with **princomp()** on neutral FST.

```{r}
neutral_pca <- princomp(as.matrix(Fst.neutral_mod))
```

c.) Calculate variance explained.

```{r}
variance <-neutral_pca$sdev^2 / sum(neutral_pca$sdev^2)
head(variance)
```
d.) Plot the first 2 PCA components with population labels.

```{r}
components <- as.data.frame(neutral_pca$scores[,1:6])

plot(components$Comp.1, components$Comp.2, xlab = "PC 1 - 86%",
     ylab= "PC 2 - 12%")

text(components$Comp.1, components$Comp.2, row.names(components),
     cex=0.6, pos=4)

abline(h =0, lty="dashed", untf = FALSE)
abline(v =0, lty="dashed", untf = FALSE)

```
## 6. Repeat PCA for ALL SNPs

a.) Run pairwise

```{r}
Fst.all<- pairwise.neifst(dataset_all) 
```

```{r}
Fst.all_mod <- 
  Fst.all %>% 
  as.data.frame() %>%
   mutate(across(everything(), ~replace_na(., mean(., na.rm = TRUE))))  %>%
  as.matrix()
```


b.) Run PCA with **princomp()** on neutral FST.

```{r}
all_pca <- princomp(as.matrix(Fst.neutral_mod))
```

c.) Calculate variance explained.

```{r}
variance <-all_pca$sdev^2 / sum(all_pca$sdev^2)
head(variance)
```
d.) Plot the first 2 PCA components with population labels.

```{r}
components <- as.data.frame(all_pca$scores[,1:6])

plot(components$Comp.1, components$Comp.2, xlab = "PC 1 - 86%",
     ylab= "PC 2 - 12%")

text(components$Comp.1, components$Comp.2, row.names(components),
     cex=0.6, pos=4)

abline(h =0, lty="dashed", untf = FALSE)
abline(v =0, lty="dashed", untf = FALSE)

```
